transactions/rbf.ts

```diff
@@ -0,0 +1,435 @@
+import * as bip39 from 'bip39';
+import EsploraProvider from '../api/esplora/esploraAPiProvider';
+import { signLedgerPSBT } from '../ledger/psbt';
+import { Transport } from '../ledger/types';
+import SeedVault from '../seedVault';
+import { AccountType, NetworkType } from '../types';
+import { RecommendedFeeResponse, Transaction, UTXO } from '../types/api/esplora';
+import { bip32 } from '../utils/bip32';
+import { getBitcoinDerivationPath, getTaprootDerivationPath } from '../wallet';
+
+// TODO: A lot of the below can be done much more easily with the consolidation logic
+// TODO: so we should refactor this file to use that once merged
+
+const areByteArraysEqual = (a: undefined | Uint8Array, b: undefined | Uint8Array): boolean => {
+  if (!a || !b || a.length !== b.length) {
+    return false;
+  }
+
+  return a.every((v, i) => v === b[i]);
+};
+
+const getRbfTransactionSummary = (transaction: Transaction) => {
+  const transactionVSize = transaction.weight / 4;
+
+  const currentTransactionInputTotals = transaction.vin.reduce((total, input) => total + input.prevout.value, 0);
+  const currentTransactionOutputTotals = transaction.vout.reduce((total, output) => total + output.value, 0);
+
+  const currentFee = currentTransactionInputTotals - currentTransactionOutputTotals;
+  const currentFeeRate = Math.ceil(currentFee / transactionVSize);
+
+  const minimumRbfFee = Math.ceil(transaction.fee + transactionVSize);
+  const minimumRbfFeeRate = Math.ceil(minimumRbfFee / transactionVSize);
+
+  return { currentFee, currentFeeRate, minimumRbfFee, minimumRbfFeeRate };
+};
+
+const isTransactionRbfEnabled = (transaction: Transaction) => {
+  if (transaction.status.confirmed) {
+    return false;
+  }
+
+  return transaction.vin.some((input) => input.sequence < 0xffffffff - 1);
+};
+
+type RBFProps = {
+  btcAddress: string;
+  ordinalsAddress: string;
+  btcPublicKey: string;
+  ordinalsPublicKey: string;
+  seedVault: SeedVault;
+  addressIndex: number;
+  network: NetworkType;
+  accountType: AccountType;
+};
+
+type TierFees = {
+  enoughFunds: boolean;
+  fee?: number;
+  feeRate: number;
+};
+
+type SoftwareCompileOptions = {
+  feeRate: number;
+};
+
+type LedgerCompileOptions = SoftwareCompileOptions & {
+  ledgerTransport: Transport;
+};
+
+type RbfRecommendedFees = {
+  medium?: TierFees;
+  high?: TierFees;
+  higher?: TierFees;
+  highest?: TierFees;
+};
+
+type InstanceCompileOptions<T> = T extends { accountType: 'software' } ? SoftwareCompileOptions : LedgerCompileOptions;
+
+class RbfTransaction<P extends RBFProps, O extends InstanceCompileOptions<P>> {
+  private baseTx: btc.Transaction;
+
+  private initialInputTotal!: number;
+
+  private initialOutputTotal!: number;
+
+  private transaction!: Transaction;
+
+  private p2btc!: btc.P2Ret;
+
+  private p2tr!: btc.P2TROut;
+
+  private network!: typeof btc.NETWORK;
+
+  private wallet!: RBFProps;
+
+  private paymentUtxos?: UTXO[];
+
+  constructor(transaction: Transaction, wallet: P) {
```

Comment:
We use the `BtcTransactionData` interface https://github.com/secretkeylabs/xverse-core/blob/a062f08934e8c8c2869f95d0f4bc6b92fbec3ff0/types/api/blockcypher/wallet.ts#L58 when fetching an array of txs or a single one on the extension, instead of the `Transaction` interface used in this file, it has a different `status` property and stuff. Should we change it here or adapt the extension to pass another object?

Reply:
I think we may need to keep it as is. The BtcTransactionData doesn't seem to have `weight` which is used in the RBF logic.

Reply:
I see we actually map the Transaction object to the BtcTransactionData object. I'll make the change to add the weight to BtcTransactionData and we'll be able to use it 👍